diofant/polys/rootoftools.py

```diff
@@ -299,9 +303,52 @@ def _complexes_sorted(cls, complexes):
             return []
         cache = {}
 
+        def is_re_disjoint(u, v):
+            if u.conj != v.conj:
+                return True
+            else:
+                return u.bx <= v.ax or v.bx <= u.ax
+
+        res_cache = {}
+
+        def match_same_real_part(u, v):
+            assert u.domain == v.domain
+            dom = u.domain
+            if tuple(u.f) not in res_cache:
+                res_cache[tuple(u.f)] = dmp_resultant(dmp_permute(u.f1, [1, 0], 1, dom),
+                                                      dmp_permute(u.f2, [1, 0], 1, dom),
+                                                      1, dom)
+            if tuple(v.f) not in res_cache:
+                res_cache[tuple(v.f)] = dmp_resultant(dmp_permute(v.f1, [1, 0], 1, dom),
+                                                      dmp_permute(v.f2, [1, 0], 1, dom),
+                                                      1, dom)
+            g = dmp_gcd(res_cache[tuple(u.f)], res_cache[tuple(v.f)], 0, dom)
```

Comment:
I would have expected `u.f` and `v.f` be the same. Otherwise the real parts of `u` and/or `v` might not necessarily be roots of `g`.

Reply:
I hadn't been thinking about what should be done that case. It could be possible that the two resultants have a common real root but that is not the real part of one (or any) of the roots. Perhaps the intervals should be refined separately for each root until there is only one real root left. Then I think that gcd would be ok.

Reply:
@jksuom, I did some cleanup for helper is_same_real_part(), which (hopefuly) make things more transparent and clear.  Let me know if you are still interested in review.

I think, that your objection "It could be possible that the two resultants have a common real root but that is not the real part of one (or any) of the roots." is not valid.  If it's the case -  is_same_real_part() return False.   If you disagree, could you explain what did I miss?

Reply:
I think that there should be no problem if, for each root u and v, the polynomial (in res_cache) having the real part as a root has only one root in the interval (u.ax, u.bx) resp. (v.ax, v.bx). I wasn't sure if that was guaranteed at this point. It seemed that there could be complex roots with far different imaginary part but very nearly the same real part (in the same interval), and then the gcd might catch a wrong real part. That should not happen if the real roots have been isolated beforehand.

Reply:
> I wasn't sure if that was guaranteed at this point.

@jksuom, but why this should be guaranteed here?

> and then the gcd might catch a wrong real part.

We test, 1) that  are in the intersection of real parts of the original ``u`` an ``v``.  
Then, if there some root (unique!) - 2) test that both original polynomial have one complex root with real part in the interval gcd_roots[0][0].

If there multple gcd_roots (or none) or original polynomials don't have complex roots (or have too much) - return False.  Why do you think that "wrong real parts", coming with gcd_roots will survive both tests?

Reply:
I agree that if there is a single root of gcd in the interval and each polynomial has a single root in its respective complex interval (as in 2 above), then the result True is correct. But I'm still concerned about  multiple roots. Is is correct to always return False if `len(gcd_roots) > 1`? Could it be possible that the real parts are the same and, in addition, each polynomial has another root with same but slightly different real part that happens to be in the same real interval though the imaginary parts may not be in the (ay, by) intervals of u and v?

Reply:
> Could it be possible that the real parts are the same and, in addition, each polynomial has another root with same but slightly different real part that happens to be in the same real interval

How this could be?  If there are other pair of roots with slightly different real part - we should continue refinement of u and v (by returning False) and eventually intervals for different multiplets of roots with same real part will be disjoint (as refinement process will shrink intersection of u and v).  That's why I do think that we will have len(gcd_roots) == 1 at some point (if u and v belongs to the multiplet of roots with same real part).

> to be in the same real interval though the imaginary parts may not be in the (ay, by) intervals of u and v

Why we ever should care about roots, outside of intervals for u and v?  is_same_real_part() is a decision procedure: for given isolation rectangles u and v we want to decide if these particular roots have same real part or not (in which case we must continue refinement).

Reply:
> You do not expect is_same_real_part returning False to mean that the real parts cannot possibly be same.

Yes, False just mean: continue refinement.  This process will continue until roots will be either disjoint by real part or gcd will have exactly one root.  In the later case we will check if both polys have complex root with real part in the interval of gcd's root.  If yes - we have found a pair of roots with same real part and should stop refinement.  If not, we will return False and continue refinement again: eventually real parts also will be different.

> I think that this should work correctly.

So, no issues with math you have found so far?

PS: Workaround for sympy/sympy#14738 simplifies code a lot, so I would like to include the real fix for that issue as well.  But original code for ``a==b`` case from 2c706b8fa6 is sufficiently straightforward (substitute x -> a + I*x and search for real solutions), so review of that part is less important.

Reply:
No math issues. This looks good. The name `is_same_real_part` was misleading me. Maybe it could be something like `confirm_same_real_part`.

Reply:
> Maybe it could be something like confirm_same_real_part.

Yes, that sounds better.  But I'm thinking about moving this functional from that helper function to the ComplexInterval.is_disjoint() method.  Some option could enable all this.